README.md
Slacko
======

![Slacko logo](slacko.png) A neat interface for [Slack](https://slack.com/)

[![Build Status](https://travis-ci.org/Leonidas-from-XIV/slacko.svg?branch=master)](https://travis-ci.org/Leonidas-from-XIV/slacko)

Your company got you Slack and you're all like "yeah, and now I am forced to
use a browser or what?". Fear no more, because as it turns out Slack has a REST
API that can be accessed.

Yeah, scripting Slack via `curl` sucks, so let's do something that sucks less.
Maybe a bit? How about using an existing
[integration](https://api.slack.com/community)? They are incomplete or for
sucky languages? Glad you asked, because I have *just the right thing* for you:
a neat interface to Slack via [OCaml](https://ocaml.org/)!

What do you need? OCaml, OPAM and a Slack token. All for free, no asterisks, no
hidden clauses. Life can be so easy.

Install that cow!
-----------------

Huzzah, I got you interested? Lucky you, Slacko can easily be installed via
[OCaml Package Manager (OPAM)](https://opam.ocaml.org/), so you just write

```sh
$ opam install slacko
```

Now wasn't this easy? Bet your ass it was! And you know what? It even gets
better: it also installed you a nifty little executable called `slack-notify`
which you can use to post directly to Slack *from your shell*, no programming
required at all. Actually, once compiled, not even OCaml is required anymore,
everything is self-contained.

In case you want to manually install it, you can, too. There's a number of
dependencies that you have to take care of yourself, for the list you'll have
to check the [slacko.opam](slacko.opam) file.

Then you can build Slacko by yourself:

```sh
$ make
```

How do I even use this thing?
-----------------------------

The workflow is always very simple: you get a token from Slack either manually
or my querying the OAuth API. Then you can instantly use any API method that
Slacko supports. If you want, you can look at snapshot of
[the Slacko docs](http://leonidas-from-xiv.github.io/slacko/) to see what
functions are available (hint: all of them).


Ready, set, go!
---------------

Not convinced yet? Check out this neat documentation from `slack-notify` to
help you out.

```
NAME
       slack-notify - Writes messages to slack

SYNOPSIS
       slack-notify [OPTION]... CHANNEL MSG

ARGUMENTS
       CHANNEL
           Name of the channel to post to

       MSG Message to send

OPTIONS
       --help[=FMT] (default=pager)
           Show this help in format FMT (pager, plain or groff).

       --icon-emoji=EMOJI
           emoji to use as the icon for this message. Overrides icon-url

       --icon-url=URL
           URL to an image to use as the icon for this message

       -t TOKEN, --token=TOKEN
           The Slack API access token

       -u USERNAME, --username=USERNAME
           Name of the bot in the chat
```

So what is this token thing? Once you signed up for Slack, you can get a token
for every team you are in from [the Slack API docs](https://api.slack.com/).
Click on the "get token" link and you're all set.

But I do want to code my own stuff!
-----------------------------------

You know, Slacko is a nice gal/guy. So you can just plug together whatever you
want with it, since it currently implements [100% of the Slack
API](https://api.slack.com/methods)! Well, except for the Real-Time-Messaging
part, which is on our TODO! How neat is this? And you can use it in your own
code without worrying about the license one bit, since the
[LGPL 3.0](https://www.gnu.org/licenses/lgpl.html) is liberal for re-use,
coupled with the
[OCaml linking exception](http://caml.inria.fr/pub/old_caml_site/ocaml/LICENSE.html).
For realz!

Something's amiss?
------------------

I know, I know, it is hardly believable that something could possibly be
missing from this library. But wonders happen, and if you run into one of these
cases, just shoot me a pull request.

Credit where credit's due
-------------------------

Thanks go to all the people who wrote
[CoHTTP](https://github.com/mirage/ocaml-cohttp), a real-world HTTP library
that doesn't suck and [Lwt](http://ocsigen.org/lwt/), an oddball library that
subconsciously *makes sense*. Also, the fine folks in
[#ocaml](http://irclog.whitequark.org/ocaml/) on
[freenode](https://freenode.net/).


CHANGES.md
0.15.0
------

* Better handling of timestamps (@paurkedal)
* Update user type to match what Slack returns (@Khady)
* Update OPAM metadata to 2.0 format
* Fixed brittle build

0.14.1
------

* Support additional arguments to `chat.post_message` and `chat.update`
* Adjust build process to use `dune` instead of `jbuilder`
* Replace `topkg` code with `dune-release`

0.14.0
------

* Higher precision timestamps.
* Adjust to changes in Slack APIs.
* Support for CoHTTP > 1.0.
* Ported to jbuilder, should yield easier integration with whole ecosystem,
  faster builds and less boilerplate.
* Depend on OCaml 4.04.
* Add support for releasing via `topkg`, thus adding this changelog.

0.13.0
------

* Fixes a number of previously broken endpoints.
* Adds an integration test to make sure that whatever is implemented keeps on
  working.

0.12.0
------

* Added type `chat` to abstract away from all types of channels Slack supports.
* Removed topic-is-too-long errors as the binding checks the topic length
  beforehand.
* Added `conversation` as a new type instead of string IM conversation type.
* More syntactic sugar in code, endpoint definitions now a long pipe.

0.11.0
------

* The binding now looks up User/Channel/Group IDs and rejects invalid ones.
* The message length is now validated: messages that are too long cannot be
  generated anymore.
* Adds support for some more new Slack methods:
  - `channels.archive`
  - `channels.create`
  - `channels.rename`
  - `channels.unarchive`
  - `groups.archive`
  - `groups.rename`
  - `groups.unarchive`

0.10.0
------

In this release, one of the main criticisms was addressed: the API calls are
now represented by their own types, so there are now channel types, user types
and many more. Some parameters can have only a limited number of values, these
are also represented using their own types so calling the methods with
incorrect values is impossible.

The code was updated to use the new Lwt 2.4.6 ppx macros instead of Camlp4, so
this is the minimum required release. Also, better use OCaml 4.02.0 for
improved support for ppx.

0.9.1
-----

* Added `users.info` method.
* Broken up `apierror` into a set of more relevant types per function.

0.9.0
-----

Time to get this code to the internetz!

Features:

* 100% API coverage and a handy tool to post messages to Slack.
* Also, some rudimentary docs and a neat logo made by yours truly.


test/slounit.ml
let lwt_test test_fun ctx = Lwt_main.run @@ test_fun ctx

let fake_slack_test test_fun ctx =
  lwt_test Fake_slack.with_fake_slack (fun () -> test_fun ctx)

let fake_slack_tests label tests =
  let open OUnit2 in
  label >::: List.map (fun (l, f) -> l >:: fake_slack_test f) tests


test/abbrtypes.ml
(* Slacko's public interface doesn't let us easily construct Slacko.user and
   Slacko.channel values from JSON, and it doesn't let us extract strings from
   those values at all. Therefore, we copy the record type that use these and
   skip the problematic fields. *)

(* Wrap Yojson.Safe.json so we don't have to keep providing printers for it. *)
type json = Yojson.Safe.json
[@@deriving yojson]
let pp_json fmt json = Format.pp_print_string fmt (Yojson.Safe.to_string json)

type abbr_authed_obj = {
  url: string;
  team: string;
  user: string;
  team_id: string;
  (* user_id: user; *)
} [@@deriving make, show, yojson { strict = false }]

let abbr_authed_obj (authed : Slacko.authed_obj) = {
  url = authed.Slacko.url;
  team = authed.Slacko.team;
  user = authed.Slacko.user;
  team_id = authed.Slacko.team_id;
}

type abbr_topic_obj = {
  value: string;
  (* creator: user; *)
  last_set: Timestamp.t;
} [@@deriving show, yojson { strict = false }]

let abbr_topic_obj (topic : Slacko.topic_obj) = {
  value = topic.Slacko.value;
  last_set = topic.Slacko.last_set;
}

type abbr_channel_obj = {
  (* id: channel; *)
  name: string;
  is_channel: bool;
  created: Timestamp.t;
  (* creator: user; *)
  is_archived: bool;
  is_general: bool;
  is_member: bool;
  (* members: user list; *)
  topic: abbr_topic_obj;
  purpose: abbr_topic_obj;
  last_read: Timestamp.t option [@default None];
  latest: json option [@default None];
  unread_count: int option [@default None];
  unread_count_display: int option [@default None];
  num_members: int option [@default None];
} [@@deriving show, yojson { strict = false }]

let abbr_channel_obj (chan : Slacko.channel_obj) = {
  name = chan.Slacko.name;
  is_channel = chan.Slacko.is_channel;
  created = chan.Slacko.created;
  is_archived = chan.Slacko.is_archived;
  is_general = chan.Slacko.is_general;
  is_member = chan.Slacko.is_member;
  topic = abbr_topic_obj chan.Slacko.topic;
  purpose = abbr_topic_obj chan.Slacko.purpose;
  last_read = chan.Slacko.last_read;
  latest = chan.Slacko.latest;
  unread_count = chan.Slacko.unread_count;
  unread_count_display = chan.Slacko.unread_count_display;
  num_members = chan.Slacko.num_members;
}

type abbr_channel_obj_list = abbr_channel_obj list
[@@deriving show, yojson]

type abbr_message_obj = {
  type': string [@key "type"];
  ts: Timestamp.t;
  (* user: user; *)
  text: string option;
  is_starred: bool option [@default None];
} [@@deriving show, yojson { strict = false }]

let abbr_message_obj (message : Slacko.message_obj) = {
  type' = message.Slacko.type';
  ts = message.Slacko.ts;
  text = message.Slacko.text;
  is_starred = message.Slacko.is_starred;
}

type abbr_history_obj = {
  latest: Timestamp.t option [@default None];
  messages: abbr_message_obj list;
  has_more: bool;
} [@@deriving show, yojson { strict = false }]

let abbr_history_obj (history : Slacko.history_obj) = {
  latest = history.Slacko.latest;
  messages = List.map abbr_message_obj history.Slacko.messages;
  has_more = history.Slacko.has_more;
}

type abbr_user_obj = {
  (* id: user; *)
  name: string;
  deleted: bool;
  color: string;
  real_name: string;
  tz: string option [@default None];
  tz_label: string;
  tz_offset: int;
  profile: json;
  is_admin: bool;
  is_owner: bool;
  is_primary_owner: bool;
  is_restricted: bool;
  is_ultra_restricted: bool;
  is_bot: bool;
  has_files: bool [@default false];
} [@@deriving show, yojson { strict = false } ]

let abbr_user_obj (user : Slacko.user_obj) = {
  name = user.Slacko.name;
  deleted = user.Slacko.deleted;
  color = user.Slacko.color;
  real_name = user.Slacko.real_name;
  tz = user.Slacko.tz;
  tz_label = user.Slacko.tz_label;
  tz_offset = user.Slacko.tz_offset;
  profile = user.Slacko.profile;
  is_admin = user.Slacko.is_admin;
  is_owner = user.Slacko.is_owner;
  is_primary_owner = user.Slacko.is_primary_owner;
  is_restricted = user.Slacko.is_restricted;
  is_ultra_restricted = user.Slacko.is_ultra_restricted;
  is_bot = user.Slacko.is_bot;
  has_files = user.Slacko.has_files;
}

type abbr_users_list_obj = {
  members: abbr_user_obj list
} [@@deriving of_yojson { strict = false }]

type abbr_user_obj_list = abbr_user_obj list
[@@deriving show]

let abbr_user_obj_list_of_yojson json =
  match abbr_users_list_obj_of_yojson json with
  | Result.Ok obj -> Result.Ok obj.members
  | (Result.Error _) as err -> err


type abbr_file_obj = {
  (* TODO file id type *)
  id: string;
  created: Timestamp.t;
  (* deprecated *)
  timestamp: Timestamp.t;

  name: string option [@default None];
  title: string;
  mimetype: string;
  pretty_type: string;
  (* user: user; *)

  mode: string;
  editable: bool;
  is_external: bool;
  external_type: string;

  size: int;

  (* These two are deprecated and appear to be gone. *)
  (* url: string; *)
  (* url_download: string; *)
  url_private: string;
  url_private_download: string;

  thumb_64: string option [@default None];
  thunb_80: string option [@default None];
  thumb_360: string option [@default None];
  thumb_360_gif: string option [@default None];
  thumb_360_w: int option [@default None];
  thumb_360_h: int option [@default None];

  permalink: string;
  edit_link: string option [@default None];
  preview: string option [@default None];
  preview_highlight: string option [@default None];
  lines: int option [@default None];
  lines_more: int option [@default None];

  is_public: bool;
  (*public_url_shared: ???;*)
  (* channels: channel list; *)
  (* groups: group list; *)
  (* ims: conversation list; *)
  initial_comment: json option [@default None];
  num_stars: int option [@default None];
} [@@deriving show, yojson { strict = false }]

let abbr_file_obj (file : Slacko.file_obj) = {
  id = file.Slacko.id;
  created = file.Slacko.created;
  timestamp = file.Slacko.timestamp;
  name = file.Slacko.name;
  title = file.Slacko.title;
  mimetype = file.Slacko.mimetype;
  pretty_type = file.Slacko.pretty_type;
  mode = file.Slacko.mode;
  editable = file.Slacko.editable;
  is_external = file.Slacko.is_external;
  external_type = file.Slacko.external_type;
  size = file.Slacko.size;
  url_private = file.Slacko.url_private;
  url_private_download = file.Slacko.url_private_download;
  thumb_64 = file.Slacko.thumb_64;
  thunb_80 = file.Slacko.thunb_80;
  thumb_360 = file.Slacko.thumb_360;
  thumb_360_gif = file.Slacko.thumb_360_gif;
  thumb_360_w = file.Slacko.thumb_360_w;
  thumb_360_h = file.Slacko.thumb_360_h;
  permalink = file.Slacko.permalink;
  edit_link = file.Slacko.edit_link;
  preview = file.Slacko.preview;
  preview_highlight = file.Slacko.preview_highlight;
  lines = file.Slacko.lines;
  lines_more = file.Slacko.lines_more;
  is_public = file.Slacko.is_public;
  initial_comment = file.Slacko.initial_comment;
  num_stars = file.Slacko.num_stars;
}

type abbr_paging_obj = {
  count: int;
  total: int;
  page: int;
  pages: int;
} [@@deriving show, yojson { strict = false }]

let abbr_paging_obj (paging : Slacko.paging_obj) = {
  count = paging.Slacko.count;
  total = paging.Slacko.total;
  page = paging.Slacko.page;
  pages = paging.Slacko.pages;
}

type abbr_files_list_obj = {
  files: abbr_file_obj list;
  paging: abbr_paging_obj;
} [@@deriving show, yojson { strict = false }]

let abbr_files_list_obj (files : Slacko.files_list_obj) = {
  files = List.map abbr_file_obj files.Slacko.files;
  paging = abbr_paging_obj files.Slacko.paging;
}

type abbr_group_obj = {
  (* id: group; *)
  name: string;
  is_group: bool;
  created: Timestamp.t;
  (* creator: user; *)
  is_archived: bool;
  (* members: user list; *)
  topic: abbr_topic_obj;
  purpose: abbr_topic_obj;
  is_open: bool option [@default None];
  last_read: Timestamp.t option [@default None];
  unread_count: int option [@default None];
  unread_count_display: int option [@default None];
  latest: json option [@default None];
} [@@deriving show, yojson { strict = false }]

let abbr_group_obj (group : Slacko.group_obj) = {
  name = group.Slacko.name;
  is_group = group.Slacko.is_group;
  created = group.Slacko.created;
  is_archived = group.Slacko.is_archived;
  topic = abbr_topic_obj group.Slacko.topic;
  purpose = abbr_topic_obj group.Slacko.purpose;
  is_open = group.Slacko.is_open;
  last_read = group.Slacko.last_read;
  unread_count = group.Slacko.unread_count;
  unread_count_display = group.Slacko.unread_count_display;
  latest = group.Slacko.latest;
}

type abbr_group_obj_list = abbr_group_obj list
[@@deriving show, yojson]

type abbr_im_obj = {
  id: string;
  is_im: bool;
  (* user: user; *)
  created: Timestamp.t;
  is_user_deleted: bool;
  unread_count: int option [@default None];
  unread_count_display: int option [@default None];
} [@@deriving show, yojson { strict = false }]

let abbr_im_obj (im : Slacko.im_obj) = {
  id = im.Slacko.id;
  is_im = im.Slacko.is_im;
  created = im.Slacko.created;
  is_user_deleted = im.Slacko.is_user_deleted;
  unread_count = im.Slacko.unread_count;
  unread_count_display = im.Slacko.unread_count_display;
}

type abbr_im_obj_list = abbr_im_obj list
[@@deriving show, yojson]


test/fake_slack.ml
(* A (currently pretty useless) fake slack implementation to run tests
   against. *)
open Lwt
open Cohttp_lwt_unix

(* The only "valid" token we accept. *)
let valid_token = "xoxp-testtoken"

(* The following values come from data captured by a relay proxy between slacko
   and slack.com. They need to be replaced with something more generic that
   doesn't depend on the arbitrary details of a particular slack team. *)
let ch_general = "C3UK9TS3C"
let ch_random = "C3TTWNCTA"
let ch_archivable = "C3XTJPLFL"
let ch_archived = "C3XTHDCTC"
let gr_seekrit = "G536YKXPE"
(* slacko doesn't have a lookup function for us, so we just use it directly. *)
let im_slackbot = "D3UMJU8VA"


let channels_json = Yojson.Safe.from_file "channels.json"
let new_channel_json = Yojson.Safe.from_file "new_channel.json"
let authed_json = Yojson.Safe.from_file "authed.json"
let random_history_json = Yojson.Safe.from_file "random_history.json"
let users_json = Yojson.Safe.from_file "users.json"
let files_json = Yojson.Safe.from_file "files.json"
let groups_json = Yojson.Safe.from_file "groups.json"
let seekrit_history_json = Yojson.Safe.from_file "seekrit_history.json"
let ims_json = Yojson.Safe.from_file "ims.json"
let slackbot_history_json = Yojson.Safe.from_file "slackbot_history.json"

let json_fields = function
  | `Assoc fields -> fields
  | _ -> failwith "Can't parse test json."


let reply_json ok fields =
  let body =
    `Assoc (("ok", `Bool ok) :: fields)
    |> Yojson.Safe.to_string
  in
  Server.respond_string ~status:`OK ~body ()

let reply_ok fields = reply_json true fields

let reply_err err fields = reply_json false (("error", `String err) :: fields)

let get_arg_opt arg req =
  Uri.get_query_param (Request.uri req) arg

let get_arg_default arg default req =
  match get_arg_opt arg req with
  | Some x -> x
  | None -> default

let get_arg arg req =
  match get_arg_opt arg req with
  | Some x -> x
  | None -> failwith @@ "Mandatory arg " ^ arg ^ " not given."

let check_auth f req body =
  match get_arg "token" req with
  | t when t = valid_token -> f req body
  | _ -> reply_err "invalid_auth" []

(* Request handlers *)

let bad_path req _body =
  let path = req |> Request.uri |> Uri.path in
  reply_err "unknown_method" ["req_method", `String path]

let api_test req _body =
  let args = req |> Request.uri |> Uri.query in
  let field_of_arg (k, v) = k, `String (List.hd v) in
  let fields = match args with
    | [] -> []
    | args -> ["args", `Assoc (List.map field_of_arg args)]
  in
  match Uri.get_query_param (Request.uri req) "error" with
  | None -> reply_ok fields
  | Some err -> reply_err err fields

let auth_test _req _body =
  reply_ok (json_fields authed_json)

let channels_archive req _body =
  match get_arg "channel" req with
  | ch when ch = ch_general -> reply_err "cant_archive_general" []
  | ch when ch = ch_archivable -> reply_ok []
  | ch when ch = ch_archived -> reply_err "already_archived" []
  | _ -> reply_err "channel_not_found" []

let channels_create req _body =
  match get_arg "name" req with
  | "#general" | "#random" -> reply_err "name_taken" []
  | "#new_channel" | _ -> reply_ok ["channel", new_channel_json]

let channels_history req _body =
  (* TODO: Check various filtering params. *)
  match get_arg "channel" req with
  | ch when ch = ch_random -> reply_ok (json_fields random_history_json)
  | _ -> reply_err "channel_not_found" []

let channels_list _req _body =
  (* TODO: Check exclude_archived param. *)
  reply_ok ["channels", channels_json]

let files_list _req _body =
  (* TODO: Check various filtering params. *)
  reply_ok (json_fields files_json)

let groups_list _req _body =
  (* TODO: Check exclude_archived param. *)
  reply_ok ["groups", groups_json]

let groups_history req _body =
  (* TODO: Check various filtering params. *)
  match get_arg "channel" req with
  | gr when gr = gr_seekrit -> reply_ok (json_fields seekrit_history_json)
  | _ -> reply_err "channel_not_found" []

let im_list _req _body =
  reply_ok ["ims", ims_json]

let im_history req _body =
  (* TODO: Check various filtering params. *)
  match get_arg "channel" req with
  | im when im = im_slackbot -> reply_ok (json_fields slackbot_history_json)
  | _ -> reply_err "channel_not_found" []

let users_list _req _body =
  (* TODO: Check presence param. *)
  reply_ok (json_fields users_json)

(* Dispatcher, etc. *)

let server ?(port=7357) ~stop () =
  let callback _conn req body =
    let handler = match req |> Request.uri |> Uri.path with
      | "/api/api.test" -> api_test
      | "/api/auth.test" -> check_auth auth_test
      | "/api/channels.archive" -> check_auth channels_archive
      | "/api/channels.create" -> check_auth channels_create
      | "/api/channels.history" -> check_auth channels_history
      | "/api/channels.list" -> check_auth channels_list
      | "/api/files.list" -> check_auth files_list
      | "/api/groups.history" -> check_auth groups_history
      | "/api/groups.list" -> check_auth groups_list
      | "/api/im.history" -> check_auth im_history
      | "/api/im.list" -> check_auth im_list
      | "/api/users.list" -> check_auth users_list
      | _ -> bad_path
    in
    handler req body
  in
  Server.create ~mode:(`TCP (`Port port)) ~stop (Server.make ~callback ())

let with_fake_slack f =
  let stop, wake = wait () in
  let srv = server ~stop () in
  let stop_server _result =
    wakeup wake ();
    srv
  in
  finalize f stop_server


src/lib/internal.ml
(*
* Slacko - Binding to the Slack API
* Copyright (C) 2014-2018 Marek Kubica <marek@xivilization.net>
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 3.0 of the License, or (at your option) any later version,
* with the special exception on linking described in file COPYING.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*)

let int64_pow b n =
  let rec loop b n acc =
    if n = 0 then acc else
    loop Int64.(mul b b) (n lsr 1)
         (if n land 1 = 0 then acc else Int64.(mul b acc)) in
  loop b n 1L

let timestamp_of_string x =
  let d_ps_of_intlit intlit =
    let sec = Int64.of_string intlit in
    let d = Int64.div sec 86_400L in
    let ps = Int64.(mul (rem sec 86_400L) 1_000_000_000_000L) in
    (Int64.to_int d, ps) in
  match
    match String.split_on_char '.' x with
    | [_] ->
        Ptime.Span.of_d_ps (d_ps_of_intlit x)
    | [sec_lit; subsec_lit] ->
      let (d, ps_int) = d_ps_of_intlit sec_lit in
      let ps_frac =
        if String.length subsec_lit <= 12 then
          let scale = int64_pow 10L (12 - String.length subsec_lit) in
          Int64.mul scale (Int64.of_string subsec_lit)
        else
          Int64.of_string (String.sub subsec_lit 0 12) in
      Ptime.Span.of_d_ps (d, Int64.add ps_int ps_frac)
    | _ -> None
  with
  | exception Failure _ -> None
  | None -> None
  | Some span -> Ptime.of_span span

let string_of_timestamp ts =
  let d, ps = Ptime.Span.to_d_ps (Ptime.diff ts Ptime.epoch) in
  let sec = Int64.(add (mul (of_int d) 86_400L) (div ps 1_000_000_000_000L)) in
  let subsec = Int64.(rem ps 1_000_000_000_000L) in
  Printf.sprintf "%Ld.%06Ld" sec (Int64.div subsec 1_000_000L)

let timestamp_of_yojson json =
  match
    match json with
    | `Int x -> Ptime.of_span (Ptime.Span.of_int_s x)
    | `Intlit x -> timestamp_of_string x
    | `String x -> timestamp_of_string x
    | _ -> None
  with
  | Some ts -> Result.Ok ts
  | None -> Result.Error "Couldn't parse timestamp"

let timestamp_to_yojson ts = `String (string_of_timestamp ts)


src/lib/timestamp.ml
(*
* Slacko - Binding to the Slack API
* Copyright (C) 2014-2019 Marek Kubica <marek@xivilization.net>
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 3.0 of the License, or (at your option) any later version,
* with the special exception on linking described in file COPYING.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*)

let int64_pow b n =
  let rec loop b n acc =
    if n = 0 then acc else
    loop Int64.(mul b b) (n lsr 1)
         (if n land 1 = 0 then acc else Int64.(mul b acc)) in
  loop b n 1L

type t = Ptime.t
let pp = Ptime.pp_human ~frac_s:6 ()

let of_string x =
  let d_ps_of_intlit intlit =
    let sec = Int64.of_string intlit in
    let d = Int64.div sec 86_400L in
    let ps = Int64.(mul (rem sec 86_400L) 1_000_000_000_000L) in
    (Int64.to_int d, ps) in
  match
    match String.split_on_char '.' x with
    | [_] ->
        Ptime.Span.of_d_ps (d_ps_of_intlit x)
    | [sec_lit; subsec_lit] ->
      let (d, ps_int) = d_ps_of_intlit sec_lit in
      let ps_frac =
        if String.length subsec_lit <= 12 then
          let scale = int64_pow 10L (12 - String.length subsec_lit) in
          Int64.mul scale (Int64.of_string subsec_lit)
        else
          Int64.of_string (String.sub subsec_lit 0 12) in
      Ptime.Span.of_d_ps (d, Int64.add ps_int ps_frac)
    | _ -> None
  with
  | exception Failure _ -> None
  | None -> None
  | Some span -> Ptime.of_span span

let to_string ts =
  let d, ps = Ptime.Span.to_d_ps (Ptime.diff ts Ptime.epoch) in
  let sec = Int64.(add (mul (of_int d) 86_400L) (div ps 1_000_000_000_000L)) in
  let subsec = Int64.(rem ps 1_000_000_000_000L) in
  Printf.sprintf "%Ld.%06Ld" sec (Int64.div subsec 1_000_000L)

let of_yojson json =
  match
    match json with
    | `Int x -> Ptime.of_span (Ptime.Span.of_int_s x)
    | `Intlit x -> of_string x
    | `String x -> of_string x
    | _ -> None
  with
  | Some ts -> Result.Ok ts
  | None -> Result.Error "Couldn't parse timestamp"

let to_yojson ts = `String (to_string ts)


src/cli/slack_notify.ml
(*
* slack-notify - Posts messages to Slack channels
* Copyright (C) 2014-2015 Marek Kubica <marek@xivilization.net>
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 3.0 of the License, or (at your option) any later version,
* with the special exception on linking described in file COPYING.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*)

let base_url =
  let doc = "The Slack API base URL" in
  Cmdliner.Arg.(value & opt (some string) None & info ["base-url"] ~docv:"URL" ~doc)

let token =
  let doc = "The Slack API access token" in
  Cmdliner.Arg.(required & opt (some string) None & info ["t"; "token"] ~docv:"TOKEN" ~doc)

let username =
  let doc = "Name of the bot in the chat" in
  Cmdliner.Arg.(value & opt (some string) None & info ["u"; "username"] ~docv:"USERNAME" ~doc)

let icon_url =
  let doc = "URL to an image to use as the icon for this message" in
  Cmdliner.Arg.(value & opt (some string) None & info ["icon-url"] ~docv:"URL" ~doc)

let icon_emoji =
  let doc = "Emoji to use as the icon for this message. Overrides icon-url" in
  Cmdliner.Arg.(value & opt (some string) None & info ["icon-emoji"] ~docv:"EMOJI" ~doc)

let channel =
  let doc = "Name of the channel to post to" in
  Cmdliner.Arg.(required & pos 0 (some string) None & info [] ~docv:"CHANNEL" ~doc)

let message =
  let doc = "Message to send" in
  Cmdliner.Arg.(required & pos 1 (some string) None & info [] ~docv:"MSG" ~doc)

let attachment =
  let doc = "Attachment text" in
  Cmdliner.Arg.(value & opt (some string) None & info ["attachment"] ~docv:"MSG" ~doc)

let info =
  let doc = "Posts messages to Slack" in
  Cmdliner.Term.info "slack-notify" ~doc

let execute base_url token username channel icon_url icon_emoji attachment_text msg =
  "Your token is " ^ token ^ ", the channel is " ^ channel
    ^ " and the message is '" ^ msg ^ "'."
    |> print_endline;

  let string_or_bust = function
    | `Success _ -> "Message posted"
    | `Invalid_auth -> "Invalid token"
    | `Channel_not_found -> "Channel unknown"
    | `Is_archived -> "Channel is archived"
    | `Msg_too_long -> "Message too long"
    | `Rate_limited -> "Rate limit active"
    | _ -> "Unknown error"
  in

  let open Lwt in
  let session = Slacko.start_session ?base_url token in
  let channel = Slacko.channel_of_string channel in
  let chat = Slacko.Channel channel in
  let msg = Slacko.message_of_string msg in
  let attachments =
    match attachment_text with
    | None -> None
    | Some text -> Some [Slacko.attachment ~text ()]
  in
  Slacko.chat_post_message session chat
    ?username:(username)
    ?icon_emoji:(icon_emoji)
    ?icon_url:(icon_url)
    ?attachments:(attachments)
    msg
  >|= (fun c ->
    print_endline @@ string_or_bust c)
  |> Lwt_main.run

let execute_t = Cmdliner.Term.(
    pure execute $ base_url $ token $ username $ channel $ icon_url $ icon_emoji
    $ attachment $ message)

let () =
  match Cmdliner.Term.eval (execute_t, info) with
    | `Error _ -> exit 1
    | _ -> exit 0
